docs: create labeling guide (#7634)
renovatebot/renovate: docs/development/issue_labeling.md

# Issue labeling

We try to keep issues well-classified by labeling them appropriately.
Only members/collaborators with the required rights can apply labels.

In theory any feature or bug should apply to either a platform, manager, datasource, versioning or a worker part.

## Basic knowledge about Renovate

You should know about platforms, package managers, datasources and versioning to label issues effectively.

- To learn about platforms, read the [Renovate docs on Platforms](https://docs.renovatebot.com/modules/platform/).
- To learn about managers, read the [Renovate docs on Managers](https://docs.renovatebot.com/modules/manager/).
- To learn about datasources, read the [Renovate docs on Datasources](https://docs.renovatebot.com/modules/datasource/).

> - To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).
> 

Reviewer: ```suggestion
- To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).

Most issues should have a label relating to either a platform, manager, datasource, versioning or worker topic.
```

# Issue labeling

We try to keep issues well-classified through use of labels.
Any repository collaborator can apply labels according to the below guidelines.

## Basic knowledge about Renovate

You should know about platforms, package managers, datasources and versioning to label issues effectively.

- To learn about platforms, read the [Renovate docs on Platforms](https://docs.renovatebot.com/modules/platform/).
- To learn about managers, read the [Renovate docs on Managers](https://docs.renovatebot.com/modules/manager/).
- To learn about datasources, read the [Renovate docs on Datasources](https://docs.renovatebot.com/modules/datasource/).
- To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).

Most issues should have a label relating to either a platform, manager, datasource, versioning or worker topic.

## Label categories

### Type of issue

<details>
    <summary>Type of issue</summary>

>     breaking

Reviewer: ```suggestion
```

    type:bug
    type:docs
    type:feature
    type:refactor

</details>

Use these to label the type of issue.
For example, use `type:bug` to label a bug type issue, and use `type:feature` for feature requests.
Only use `type:refactor` for code changes, don't use `type:refactor` for documentation type changes.

> Use the `breaking` label for Issues or PRs which contain changes that are not backwards compatible and require a major version bump.

Reviewer: ```suggestion
Add the `breaking` label for Issues or PRs which contain changes that are not backwards compatible and require a major version bump.
```

### Priority

<details>
    <summary>Priority</summary>

    priority-1-critical
    priority-2-important
    priority-3-normal
    priority-4-low

</details>

Use these to assign a priority level to an issue.
Make a best-effort attempt to select a proper priority.

> Nothing bad will happen if you select a "wrong" priority.

Reviewer: ```suggestion
Nothing bad will happen if you select a "wrong" priority.
At a high level: critical = needs immediate fix, important = to be prioritized ahead of others, normal = default priority, low = trivial issue, or impacts a very small % of the user base.
```

- To learn about managers, read the [Renovate docs on Managers](https://docs.renovatebot.com/modules/manager/).
- To learn about datasources, read the [Renovate docs on Datasources](https://docs.renovatebot.com/modules/datasource/).
- To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).

## Label categories

### Priority

<details>
    <summary>Priority</summary>

    pri0-blocker
    pri1-critical
    pri2-important
    pri3-normal
    pri4-low
    wontfix

</details>

Use these to assign a priority level to an issue.

QUESTION: Who gets to decide the priority level? @rarkins and maintainers, or can a triaging member select a priority level?

QUESTION: Maybe we should copy paste the priority descriptions from the labels into this doc?

### Type of issue

<details>
    <summary>Type of issue</summary>

    bug
    docs
    duplicate
    feature
    refactor
    breaking

</details>

Use these to label the type of issue.
For example, use `bug` to label a bug type issue, and use `feature` for feature requests.
Only use `refactor` for code changes, don't use `refactor` for documentation type changes.

> QUESTION: Do we use breaking for "breaking" changes in a PR, or as a "the house is on fire" marker for issues, or both?
> Use the `breaking` label for pull requests that have breaking changes.
> Use the `breaking` label for when a bug is realy major.

Reviewer: ```suggestion
Use the `breaking` label for Issues or PRs which contain changes that are not backwards compatible and require a major release bump.
```

> QUESTION: Do we even need a `duplicate` label, there's **mark as duplicate** functionality in GitHub nowadays...

Reviewer: Agreed, remove it

### Housekeeping

<details>
    <summary>Housekeeping</summary>

    good first issue
    help wanted
    needs reproduction
    reproduced

</details>

> Add a label `good first issue` to issues that are small, easy to fix, and do-able for a newcomer.

Reviewer: ```suggestion
Add a label `good first issue` to issues that are small, easy to fix, and do-able for a newcomer.
This label is sometimes picked up by tools or websites that try to encourage people to contribute to open source.
```

> Add a label `needs reproduction` if the issue can only be fixed once there's a reproduction.

Reviewer: ```suggestion
Add a label `needs reproduction` if nobody's reproduced it in a public repo yet and such a reproduction is necessary before further work can be done.
```

Add the label `reproduced` once there is a public reproduction.

> QUESTION: When/who should use the `help wanted` label?

Reviewer: I use this to indicate - usually to the issue creator - that I need them or someone else to do some work here or it is unlikely to get done.

### Miscellaneous

<details>
    <summary>Miscellaneous</summary>

>     gitfs
>     hacktoberfest
>     hacktoberfest-accepted
>     internal
>     ready

Reviewer: ```suggestion
```

These can all be removed

    self hosted

</details>

QUESTION: What is the `gitfs` label about?
QUESTION: Should we document the `hacktoberfest` labels?
QUESTION: What is the `internal` label about?
QUESTION: Do we even use the `ready` label now Renovate has a project board?

> QUESTION: I think `self hosted` belongs in the `platform:` type label section instead???

Reviewer: Any of the platforms can be self-hosted, while both GitHub and GitLab can be hosted. I use this label to identify when an issue is applicable only to users who self-administer their own bot.

### Datasource

<details>
    <summary>Datasource</summary>

    datasource:docker
    datasource:git-submodule
    datasource:git-labels
    datasource:jenkins
    datasource:maven
    datasource:nuget
    datasource:packagist
    datasource:pypi
    datasource:rubygems
    datasource:terraform-module
    datasource:terraform-provider

</details>

> QUESTION: When should somebody use a datasource label?

Reviewer: When it is applicable specifically to particular datasources (e.g. as defined in the docs list of datasources)

    manager:dockerfile
    manager:github-actions
    manager:gitlab-ci
    manager:gomod
    manager:gradle
    manager:helm
    manager:helm-values
    manager:kubernetes
    manager:kustomize
    manager:maven
    manager:meteor
    manager:mix
    manager:npm
    manager:nuget
    manager:pip_requirements
    manager:pip_setup
    manager:pipenv
    manager:poetry
    manager:ruby-version
    manager:sbt
    manager:swift
    manager:terraform
    manager:travis

</details>

"manager" is short for "package manager".
Add the relevant manager labels to the issue.
If there are multiple managers affected, add labels for all of them.

### New stuff

<details>
    <summary>New stuff</summary>

    new datasource
    new package manager
    new platform
    new versioning

</details>

Apply these labels when somebody opens a `feature` type issue requesting a new datasource, package manager, platform, or new versioning scheme.

### Platform

<details>
    <summary>Platform labels</summary>

    platform:azure
    platform:bitbucket
    platform:gitea
    platform:github
    platform:gitlab

</details>

Use these to mark the platform that is affected by this issue.

> QUESTION: Should the `self hosted` label be in this category as well? Or are we missing a `platform:self-hosted` label?

Reviewer: No, because it could be self-hosted *and* any of the above.

### Worker

<details>
    <summary>Worker</summary>

    worker:branch
    worker:global
    worker:onboarding
    worker:pr

</details>

> QUESTION: I have no clue what a worker even is....

Reviewer: Worker is kind of the "core" logic of Renovate. I think it's a good idea to have these, or they haven't been used heavily so far. In theory any feature or bug should apply to either a platform, manager, datasource, versioning or a worker part.

### Semantic release bot

<details>
    <summary>Semantic release bot</summary>

    released
    semantic-release

</details>

The semantic release bot applies the `released` label whenever the fix for an issue is packaged into a new release.

> QUESTION: I don't know what the `semantic-release` label is about.

Reviewer: I don't know either. Also, I would be fine to stop using `released` as a label because semantic-release already comments in every issue and PR anyway.

Author: I think you'll need to edit some `semantic` bot configuration to get it to stop using the label?

Reviewer: See https://github.com/renovatebot/renovate/pull/7654

Author: I've removed the entire section now that you've merged #7654.

# Issue labeling

We try to keep issues well-classified by labeling them appropriately.
Only members/collaborators with the required rights can apply labels.

## Basic knowledge about Renovate

You should know about platforms, package managers, datasources and versioning to label issues effectively.

- To learn about platforms, read the [Renovate docs on Platforms](https://docs.renovatebot.com/modules/platform/).
- To learn about managers, read the [Renovate docs on Managers](https://docs.renovatebot.com/modules/manager/).
- To learn about datasources, read the [Renovate docs on Datasources](https://docs.renovatebot.com/modules/datasource/).
- To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).

## Label categories

### Severity

<details>
    <summary>Severity</summary>

    breaking
    major

</details>

> QUESTION: When/who should use these severity labels?

Reviewer: Breaking is added if we determine it's a backwards-incompatible change that requires us to publish a new major release.

### Priority

<details>
    <summary>Priority</summary>

>     pri0-blocker

Reviewer: Remove blocker, combine with critical instead

Reviewer: I've removed it from the repo

    pri1-critical
    pri2-important
    pri3-normal
    pri4-low
    wontfix

</details>

Use these to assign a priority level to an issue.

> QUESTION: Who gets to decide the priority level? @rarkins and maintainers, or can a triaging member select a priority level?

Reviewer: Any collaborator in the repo with rights can assign any label

Author: I know that any collaborator with the required rights can assign labels, the question I'm really asking here is: "Who decides what priority an issue is? Is a best-effort guess from any collaborator okay?"

I hope I'm making sense here... :smile:

Reviewer: Yes, that is ok. Nothing bad happens if someone has a different opinion on priority.

> QUESTION: Maybe we should copy paste the priority descriptions from the labels into this doc?

Reviewer: This is an important field so we could start by copying but best if we expand it.

Reviewer: I've added descriptions to those which were missing

We try to keep issues well-classified by labeling them appropriately.
Only members/collaborators with the required rights can apply labels.

## Basic knowledge about Renovate

You should know about platforms, package managers, datasources and versioning to label issues effectively.

- To learn about platforms, read the [Renovate docs on Platforms](https://docs.renovatebot.com/modules/platform/).
- To learn about managers, read the [Renovate docs on Managers](https://docs.renovatebot.com/modules/manager/).
- To learn about datasources, read the [Renovate docs on Datasources](https://docs.renovatebot.com/modules/datasource/).
- To learn more about versioning, read the [Renovate docs on Versioning](https://docs.renovatebot.com/modules/versioning/).

## Label categories

### Severity

<details>
    <summary>Severity</summary>

    breaking

>     major

Reviewer: I removed `major` now and I'm not sure that `breaking` then needs its own category. It can maybe be related to feature/fix/refactor ?

Author: I've removed the `major` label, and moved the breaking label to the Type of issue section.
I'm not sure what `breaking` denotes though, see the new file for questions. :smile:
